zokrates_ast/src/ir/witness.rs

```diff
@@ -41,54 +41,78 @@ impl<T: Field> Witness<T> {
         Witness(BTreeMap::new())
     }
 
-    pub fn write<W: Write>(&self, writer: W) -> io::Result<()> {
-        let mut wtr = csv::WriterBuilder::new()
-            .delimiter(b' ')
-            .flexible(true)
-            .has_headers(false)
-            .from_writer(writer);
-
-        // Write each line of the witness to the file
+    pub fn write<W: Write>(&self, mut writer: W) -> io::Result<()> {
+        let length = self.0.len();
+        writer.write_all(&length.to_le_bytes())?;
+
         for (variable, value) in &self.0 {
-            wtr.serialize((variable.to_string(), value.to_dec_string()))?;
+            variable.write(&mut writer)?;
+            value.write(&mut writer)?;
         }
-
         Ok(())
     }
 
     pub fn read<R: Read>(mut reader: R) -> io::Result<Self> {
-        let mut rdr = csv::ReaderBuilder::new()
-            .delimiter(b' ')
-            .flexible(true)
-            .has_headers(false)
-            .from_reader(&mut reader);
-
-        let map = rdr
-            .deserialize::<(String, String)>()
-            .map(|r| {
-                r.map(|(variable, value)| {
-                    let variable = Variable::try_from_human_readable(&variable).map_err(|why| {
-                        io::Error::new(
-                            io::ErrorKind::Other,
-                            format!("Invalid variable in witness: {}", why),
-                        )
-                    })?;
-                    let value = T::try_from_dec_str(&value).map_err(|_| {
+        let mut witness = Self::empty();
+
+        let mut buf = [0; std::mem::size_of::<usize>()];
+        reader.read_exact(&mut buf)?;
+
+        let length: usize = usize::from_le_bytes(buf);
+
+        for _ in 0..length {
+            let var = Variable::read(&mut reader)?;
+            let val = T::read(&mut reader)?;
+
+            witness.insert(var, val);
+        }
+
+        Ok(witness)
+    }
+
+    pub fn write_json<W: Write>(&self, writer: W) -> io::Result<()> {
+        let map = self
+            .0
+            .iter()
+            .map(|(k, v)| (k.to_string(), serde_json::json!(v.to_dec_string())))
+            .collect::<serde_json::Map<String, serde_json::Value>>();
+
+        serde_json::to_writer_pretty(writer, &map)?;
+        Ok(())
+    }
+
+    pub fn read_json<R: Read>(reader: R) -> io::Result<Self> {
```

Comment:
If the json is just for debugging purposes I don't think we should give the option to read it

Reply:
We do need this in the integration test, check `integration.rs`. We only check the output variables in witness (which I assume is because other variables could change). This way we can still easily specify parts of a witness for the tests

Reply:
I see, then all the relevant code should be moved to the tests as helper functions or tagged with `#[cfg(test)]`